### PR TITLE
Support multiple daily process locking backends with distributed polling

### PR DESCRIPTION
**Pull request by murrant**
Using Laravel locks allows us to use redis or memcached locks for daily.sh in conjunction with distributed polling.

Uses database for locking by default.  To use redis or memcached, set this in your .env file:
`CACHE_DRIVER=redis`
or
`CACHE_DRIVER=memcached`

Please give a short description what your pull request is for

DO NOT DELETE THE UNDERLYING TEXT

#### Please note

> Please read this information carefully. You can run `./scripts/pre-commit.php` to check your code before submitting.

- [x] Have you followed our [code guidelines?](http://docs.librenms.org/Developing/Code-Guidelines/)
- [x] If my Pull Request does some changes/fixes/enhancements in the WebUI, I have inserted a screenshot of it.

#### Testers

If you would like to test this pull request then please run: `./scripts/github-apply <pr_id>`, i.e `./scripts/github-apply 5926`
After you are done testing, you can remove the changes with `./scripts/github-remove`.  If there are schema changes, you can ask on discord how to revert.

### CODE REVIEW

File: includes/sql-schema/update.php [hunk 1]
```diff
@@ -108,10 +96,5 @@
         echo Artisan::output();
     }
 
-    if (isset($schemaLock)) {
-        $schemaLock->release();
-    }
-} catch (LockException $e) {
-    echo $e->getMessage() . PHP_EOL;
-    $return = 1;
+    $schemaLock->release();
```
Comment (Jellyfrog) [inline includes/sql-schema/update.php:99]: Wouldn't this one be released even if skip =true?

File: alerts.php [hunk 2]
```diff
@@ -27,33 +27,34 @@
  * @author: Heath Barnhart <hbarnhart@kanren.net>
  */
 
-use LibreNMS\Util\FileLock;
 use LibreNMS\Alert\RunAlerts;
+use LibreNMS\Util\FileLock;
```
Comment (Jellyfrog) [inline alerts.php:31]: Typo?

File: includes/sql-schema/update.php (hunk 1 above)
Comment (murrant) [reply includes/sql-schema/update.php:99]: You can't release a lock you don't own unless you forceRelease()

File: alerts.php (hunk 2 above)
Comment (murrant) [reply alerts.php]: Legacy, I was looking at refactoring the existing locks, but went this route instead.  Thanks, removed this and one other reference.

**Comment by Jellyfrog** [comment]
If I understand this right, you're implementing Laravels lock feature for the `file`-driver ?  
Couldn't we utilize Laravel 7's support for `database` instead? Seems like a fair trade-of if the user dosn't have redis/memcache?  https://laravel.com/docs/7.x/cache#atomic-locks

**Comment by Jellyfrog** [comment]
Other than that, it looks nice

**Comment by murrant** [comment]
@Jellyfrog haha, didn't see they added a database driver for 7.  That would be a good compromise.  I was worried about what to do with distributed poller setups as we have to set the cache driver to something that supports locking by default.

**Comment by murrant** [comment]
@librenms/reviewers Ready for review/merge

File: app/Providers/AppServiceProvider.php [hunk 3]
```diff
@@ -156,4 +159,17 @@ private function bootCustomValidators()
...
     }
+
+    protected function extendFileCacheDriver(): void
+    {
+        Cache::extend('file', function ($app) {
+            return Cache::repository(
+                new LockingFileStore(
+                    $app['files'],
+                    $app['config']->get('cache.stores.file.path'),
+                    $app['config']->get('cache.stores.file.permissions')
```
Comment (Jellyfrog) [inline app/Providers/AppServiceProvider.php:170]: Doesn't exist?

**Comment by Jellyfrog** [comment]
We're keeping the custom file locking but not using it if I'm reading this correct? Else we're missing docs

**Comment by murrant** [comment]
I wasn't sure about the custom file locks as it is no longer using it.  I think I can rip it out.
The only reason I left it in was someone might have file set for cache, probably not a good reason and the code will sit mostly unused and bitrot.

File: alerts.php [hunk 4]
```diff
@@ -28,32 +28,30 @@
...
-    echo 'ClearStaleAlerts():' . PHP_EOL;
-    $alerts->clearStaleAlerts();
-    echo "RunFollowUp():\r\n";
-    $alerts->runFollowUp();
-    echo "RunAlerts():\r\n";
-    $alerts->runAlerts();
-    echo "RunAcks():\r\n";
-    $alerts->runAcks();
-    echo 'End  : ' . date('r') . "\r\n";
+$alerts_lock = Cache::lock('alerts');
```
Comment (Jellyfrog) [inline alerts.php:41]: Missing use Cache?

File: daily.php [hunk 5]
```diff
@@ -11,8 +11,6 @@
 use Illuminate\Database\Eloquent\Collection;
 use LibreNMS\Alert\AlertDB;
 use LibreNMS\Config;
-use LibreNMS\Exceptions\LockException;
-use LibreNMS\Util\MemcacheLock;
```
Comment (Jellyfrog) [inline daily.php:15]: Missing use Cache?

File: alerts.php (hunk 4 above)
Comment (murrant) [reply alerts.php:41]: hehe, the ide doesn't like the use statement...  leaving as-is.

**Comment by murrant** [comment]
@librenms/reviewers This seems ready to merge to me, any objections?

**Comment by hanserasmus** [comment]
I have tested both SQL locks and Redis locks on a CentOS 8 distributed system. One node running php 7.3 from Appstream repo and the other php74 from Remi. It worked well. No errors. Great job!

**Comment by f0o** [comment]
LGTM 😅

**Comment by murrant** [comment]
This pull request has been mentioned on **LibreNMS Community**. There might be relevant details there:

https://community.librenms.org/t/v1-69-release-changelog-october-2020/13837/1

**Comment by murrant** [comment]
This pull request has been mentioned on **LibreNMS Community**. There might be relevant details there:

https://community.librenms.org/t/mariadb-galera-cluster-general-error-1205-lock-wait-timeout-exceeded/14201/2

**Pull request merged**